### ISSUE TITLE
multi-label visual distribution by t-SNE

### ISSUE THREAD

**Issue by HerbertChan** [GenjiB/CM-Co-Occurrence-AVVP#1]
Hello, how did you visualize t-SNE for this multi-label situation? Did you make a dictionary like figure-4 out of the multi-hot labels?
Could you share the file where you did the multi-label visualization using t-SNE? It seems the mytsne.py doesn't work for me.

**Comment by GenjiB** [comment]
that does not work for me either. I finally got the figure with sklearn-tsne in CPU, which is pretty slow.

**Comment by HerbertChan** [comment]
这是来自QQ邮箱的假期自动回复邮件。你好，我最近正在休假中，无法亲自回复你的邮件。我将在假期结束后，尽快给你回复。

**Comment by HerbertChan** [comment]
Thank you for your response.
I reduced the features to two (12000, ) vectors using a third-party package tsnecuda. 
Still, I have a question, figure 3 in the paper gives 133 different label combinations, so how should I set the parameters in the **plt.scatter()**  function in order to get a similar figure? Is it necessary to make a new label list of 133 classes as the incoming value of **color**? I'd like to know how you deal with it.

**Comment by GenjiB** [comment]
`Yes, you have to determine all colors for all combinations. I have some toy examples. You can check this out

    from sklearn.manifold import TSNE

	import matplotlib.pyplot as plt

	f_a_all = []
	f_v_all = []
	total_label = []
	import matplotlib.cm as cm

	
	
	for batch_idx, sample in enumerate(val_loader):
		
		# audio_spec, gt = sample['audio_spec'].to('cuda'), sample['GT'].to('cuda')
		audio_spec, gt= sample['audio_spec'].to('cuda'), sample['GT'].to('cuda')
		image = sample['image'].to('cuda')
		audio_vgg = sample['audio_vgg'].float().to('cuda')

		with torch.no_grad():
			f_a, f_v  = model([audio_spec, audio_vgg], image)

			gt = rearrange(gt, 'b t class -> (b t) class') #.argmax(dim=-1)
			total_label.append(gt[gt.argmax(dim=-1)!=28])

			f_a_all.append(f_a.squeeze(1)[gt.argmax(dim=-1)!=28])
			f_v_all.append(f_v.squeeze(1)[gt.argmax(dim=-1)!=28])

	

	f_a_all = torch.vstack(f_a_all)

	f_v_all = torch.vstack(f_v_all)
	total_label = torch.vstack(total_label)
	


	all_f = torch.cat((f_a_all, f_v_all), dim=0)
	
	# X_embedded = TSNE(n_components=2, perplexity=30.0, n_iter=250, verbose=True).fit_transform(f_v_all[:100].cpu())

	print('computing tsne')
	Y = TSNE(n_components=2).fit_transform(all_f.cpu().numpy())
	Y_min, Y_max = Y.min(0), Y.max(0)
	Y_norm = (Y- Y_min)/(Y_max - Y_min)


	colors = cm.rainbow(np.linspace(0, 1, 28))
	plt.figure(figsize=(8,8))
	
	
	plt.scatter(Y_norm[:3305,0], Y_norm[:3305,1], marker='+', color=colors[total_label.argmax(dim=-1).cpu()],s=20,linewidths=0.1)
	plt.scatter(Y_norm[3305:,0], Y_norm[3305:,1], marker='.', color=colors[total_label.argmax(dim=-1).cpu()],s=20,linewidths=0.1)



	plt.scatter(Y_norm[3305,0], Y_norm[3305,1], label='audio', marker='+', color='black',s=0,linewidths=0.1)
	plt.scatter(Y_norm[99,0], Y_norm[99,1], label='vision', marker='.', color='black',s=0,linewidths=0.1)

	# plt.scatter(Y_norm[:4020,0], Y_norm[:4020,1], label='audio', color='blue',s=1)
	# plt.scatter(Y_norm[4020:,0], Y_norm[4020:,1], label='vision', color='red', s=1)
	# plt.legend(loc="upper right", scatterpoints=1, fontsize=10)
	lgnd = plt.legend(loc="upper right", numpoints=1, fontsize=10)
	lgnd.legendHandles[0]._sizes = [30]
	lgnd.legendHandles[1]._sizes = [30]

	plt.xticks([])
	plt.yticks([])
	plt.savefig("ViT-L-labels_HR.pdf")`
